### PR TITLE
Refactor implementation of pure-pursuit controller

### PR DESCRIPTION
**Pull request by shadaj**

### CODE REVIEW

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala [hunk 1]
```diff
@@ -165,5 +163,70 @@ trait PurePursuitControllers {
       }
     )
   }
+
+  def followWayPointsController(wayPoints: Seq[Point], position: PeriodicSignal[Point])
+                               (implicit hardware: UnicycleHardware,
+                                signal: Signal[UnicycleProperties]): (PeriodicSignal[UnicycleSignal], Signal[Option[Length]]) = {
+    var previousLookAheadPoint: Option[Point] = None
+
+    val biSegmentPaths = wayPoints.sliding(3).map { points =>
```
Comment (PhilipAxelrod) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala:172]: You should prepend the initial position

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala [hunk 2]
```diff
@@ -165,5 +163,70 @@ trait PurePursuitControllers {
...
+        if (points.size == 3) {
+          Some(Segment(points(1), points(2)))
+        } else {
+          None
+        }
+      )
+    }
+
+    var currPath: (Segment, Option[Segment]) = (
+      Segment(wayPoints(0), wayPoints(1)),
```
Comment (PhilipAxelrod) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala:184]: Doesn't handle case of only one way point. Prepend initial position to fix this.

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala [hunk 3]
```diff
@@ -165,5 +163,70 @@ trait PurePursuitControllers {
...
+    val historyUpdatingLookAheadPoint = lookAheadPoint.withCheck{p =>
+      previousLookAheadPoint = Some(p)
+    }
+
+    val forwardOutput = pointDistanceControl(
+      position,
+      historyUpdatingLookAheadPoint)._1
+    val distanceToLast = position.map(_ distanceTo wayPoints.last)
+
+    val errorToLast = distanceToLast.peek.map { d =>
```
Comment (PhilipAxelrod) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala:218]: This breaks  down if after traveling some path, we need to return to the starting point. Instead use the distance to the look ahead point

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitTasks.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+      val position = XYPosition(
+        hardware.turnPosition.map(_ - initialTurnPosition).map(compassToTrigonometric),
+        hardware.forwardPosition
+      )
+
+      val (unicycle, error) = followWayPointsController(wayPoints, position)
+
+      drive.setController(lowerLevelOpenLoop(unicycle.withCheck { _ =>
+        if (error.get.forall(_ < tolerance)) {
```
Comment (Paxelord) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitTasks.scala]: Why is error an option?

**Review by PhilipAxelrod — CHANGES_REQUESTED** [review]

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala (hunk 2 above)
Comment (PhilipAxelrod) [reply commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala:184]: You'll get a indexOutOfBounds exception

**Review by PhilipAxelrod — CHANGES_REQUESTED** [review]

**Review by PhilipAxelrod — APPROVED** [review]
Sure, lgtm

**Pull request merged**